dyno-core/src/main/java/com/netflix/dyno/connectionpool/impl/lb/TokenAwareSelection.java

```diff
@@ -32,77 +34,95 @@
-	
-	public TokenAwareSelection() {
-		
-		this.tokenMapper = new BinarySearchTokenMapper(new Murmur1HashPartitioner());
-	}
-
-	@Override
-	public void initWithHosts(Map<HostToken, HostConnectionPool<CL>> hPools) {
-		
-		tokenPools.putAll(CollectionUtils.transformMapKeys(hPools, new Transform<HostToken, Long>() {
-
-			@Override
-			public Long get(HostToken x) {
-				return x.getToken();
-			}
-			
-		}));
-
-		this.tokenMapper.initSearchMecahnism(hPools.keySet());
-	}
-
-	@Override
-	public HostConnectionPool<CL> getPoolForOperation(BaseOperation<CL, ?> op) throws NoAvailableHostsException {
-		
-		String key = op.getKey();
-		HostToken hToken = this.getTokenForKey(key);
-		
-		HostConnectionPool<CL> hostPool = null;
-		if (hToken != null) {
-			hostPool = tokenPools.get(hToken.getToken());
-		}
-		
-		if (hostPool == null) {
-			throw new NoAvailableHostsException("Could not find host connection pool for key: " + key + ", hash: " +
-                    tokenMapper.hash(key));
-		}
-		
-		return hostPool;
-	}
-
-	@Override
-	public Map<HostConnectionPool<CL>,BaseOperation<CL,?>> getPoolsForOperationBatch(Collection<BaseOperation<CL, ?>> ops) throws NoAvailableHostsException {
-		throw new RuntimeException("Not Implemented");
-	}
-	
-	@Override
-	public List<HostConnectionPool<CL>> getOrderedHostPools() {
-		return new ArrayList<HostConnectionPool<CL>>(tokenPools.values());
-	}
-	
-	@Override
-	public HostConnectionPool<CL> getPoolForToken(Long token) {
-		return tokenPools.get(token);
-	}
-	
-	public List<HostConnectionPool<CL>> getPoolsForTokens(Long start, Long end) {
-		throw new RuntimeException("Not Implemented");
-	}
+    private final BinarySearchTokenMapper tokenMapper;
+
+    private final ConcurrentHashMap<Long, HostConnectionPool<CL>> tokenPools = new ConcurrentHashMap<Long, HostConnectionPool<CL>>();
+
+    public TokenAwareSelection() {
+
+        this.tokenMapper = new BinarySearchTokenMapper(new Murmur1HashPartitioner());
+    }
+
+    @Override
+    public void initWithHosts(Map<HostToken, HostConnectionPool<CL>> hPools) {
+
+        tokenPools.putAll(CollectionUtils.transformMapKeys(hPools, new Transform<HostToken, Long>() {
+
+            @Override
+            public Long get(HostToken x) {
+                return x.getToken();
+            }
+
+        }));
+
+        this.tokenMapper.initSearchMecahnism(hPools.keySet());
+    }
+
+    /**
+     * If a hashtag is provided by Dynomite then we use that to craete the key to hash.
+     */
+    @Override
+    public HostConnectionPool<CL> getPoolForOperation(BaseOperation<CL, ?> op, String hashtag) throws NoAvailableHostsException {
+
+        String key = op.getKey();
+
+        HostToken hToken = null;
+        if (hashtag == null) {
+            hToken = this.getTokenForKey(key);
+        } else {            
+            String hashValue = StringUtils.substringBetween(key,Character.toString(hashtag.charAt(0)), Character.toString(hashtag.charAt(1)));
```

Comment:
@robzienert @shailesh33 @jcacciatore IF there is a hashtag defined, we use the substring inside the hashtag to determine the key to use for getting the node with the proper token.

Reply:
as discussed verify substringBetween does the right thing for us